classes/loginflow/authcode.php

```diff
@@ -617,13 +617,16 @@ protected function handlelogin(string $oidcuniqid, array $authparams, array $tok
             }
             $username = trim(\core_text::strtolower($username));
             $tokenrec = $this->createtoken($oidcuniqid, $username, $authparams, $tokenparams, $idtoken, 0, $originalupn);
-
-            $existinguserparams = ['username' => $username, 'mnethostid' => $CFG->mnet_localhost_id];
+            $userinfo = $this->get_userinfo($username);
+            if (!$CFG->allowaccountssameemail && array_key_exists('email', $userinfo)) {
+                $existinguserparams = ['email' => $userinfo['email'], 'mnethostid' => $CFG->mnet_localhost_id];
+            } else {
+                $existinguserparams = ['username' => $username, 'mnethostid' => $CFG->mnet_localhost_id];
+            }
```

Comment:
I prefer not to have and if-else block, but it might be required here. Also because the email is not set in the alternative version, which means it would read as:

```suggestion
            $existinguserparams = ['mnethostid' => $CFG->mnet_localhost_id];
            if (!$CFG->allowaccountssameemail && array_key_exists('email', $userinfo)) {
                $existinguserparams['email'] = $userinfo['email'];
            } else {
                $existinguserparams['username'] = $username;
            }
```

Would this alternative read more easily?

Reply:
In this case I think the conditional assignment reads just a tad cleaner than the conditional addition, because it isnt too heavy, just 2 elements. Im opting to leave as is